### PR TITLE
WT-9319 Save the user provided cursor bounds to the cursor

### PR DESCRIPTION
**Pull request by jiechenbo**
This ticket implements the initial iteration of range cursor set bounds.

### CODE REVIEW

**Comment by jiechenbo** [comment]
There is one edge case that I'm not sure if we need to test here, where the `inclusive` option is false, and in the case that 
```
cursor->set_key("D")
cursor->bound("bound=lower,inclusive=false")
cursor->set_key("D")
cursor->bound("bound=upper,inclusive=false")
```
Should the cursor bound on the second call return an ERROR here? Thinking its not possible to find something that is > D and < D.

File: src/cursor/cur_std.c [hunk 1]
```diff
@@ -1175,11 +1177,77 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+                  session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
+                if (exact < 0)
+                    WT_ERR_MSG(session, EINVAL,
+                      "The upper bounds is not lexicographically greater than the lower bound");
+            }
+            F_SET(cursor, WT_CURSTD_BOUND_UPPER | WT_CURSTD_BOUND_UPPER_INCLUSIVE);
+            WT_ERR(__wt_buf_set(session, &cursor->upper_bound, key.data, key.size));
+
+            /* Clear the key the cursor was holding. */
+            WT_CLEAR(cursor->key);
```
Comment (ddanderson) [inline src/cursor/cur_std.c:1202]: I think this WT_CLEAR may leak memory.  If you really want to get rid of the key's allocated memory, you can call     `__wt_buf_free(session, &cursor->key);` and then clear.  But is there a reason the clear the key?  Does the doc say that we do?

File: src/cursor/cur_std.c [hunk 2]
```diff
@@ -1175,11 +1177,77 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+                  session, CUR2BT(cursor)->collator, &key, &cursor->upper_bound, &exact));
+                if (exact > 0)
+                    WT_ERR_MSG(session, EINVAL,
+                      "The lower bounds is not lexicographically less than the upper bound");
+            }
+            F_SET(cursor, WT_CURSTD_BOUND_LOWER | WT_CURSTD_BOUND_LOWER_INCLUSIVE);
+            WT_ERR(__wt_buf_set(session, &cursor->lower_bound, key.data, key.size));
+
+            /* Clear the key the cursor was holding. */
+            WT_CLEAR(cursor->key);
```
Comment (ddanderson) [inline src/cursor/cur_std.c:1221]: Same problem here.

**Review by ddanderson — COMMENTED** [review]
Noted a potential problem in clearing the key.

On more thing, if you could fix PR comment.  "WT-9319 Investigate range" -- it's more than an investigation, the comment should closely match the JIRA ticket.  It will make for better explanation in the git comments.

File: src/cursor/cur_std.c (hunk 1 above)
Comment (jiechenbo) [reply src/cursor/cur_std.c]: Brief looking at the technical docs it doesn't say that we clear the key when we set bounds. Talking with @luke-pearson we have decided free the key. I can see benefits of both methods here. If we clear the key, it provides a clear understanding next(), prev() or search() calls will behave as appriopately. I am talking about the case that if the user provides an upper bound, then performs a next(). If we do not clear the key, naturally the cursor will return WT_NOTFOUND, i am not sure if this can cause confusion to the user. @luke-pearson  what are your thoughts here?

File: src/cursor/cur_std.c [hunk 3]
```diff
@@ -1175,13 +1176,83 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
         WT_ERR(__wt_config_gets(session, cfg, "bound", &cval));
         if (cval.len == 0)
             WT_ERR_MSG(session, EINVAL, "setting bounds must require the bound configuration set");
+
+        /* The cursor must have a key set to place the lower or upper bound. */
+        if (!F_ISSET(cursor, WT_CURSTD_KEY_SET))
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1181]: You should use: __cursor_checkkey here instead it handles the error messaging.

File: src/cursor/cur_std.c [hunk 4]
```diff
@@ -1175,13 +1176,83 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+            WT_ERR_MSG(session, EINVAL, "setting bounds must require the key to be set");
+
+        if (WT_STRING_MATCH("upper", cval.str, cval.len)) {
+            /*
+             * If the lower bounds are set, make sure that the upper bound is greater than the
+             lower
+             * bound.
+             */
+            WT_ERR(__wt_cursor_get_raw_key(cursor, &key));
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_LOWER | WT_CURSTD_BOUND_LOWER_INCLUSIVE)) {
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1191]: Why are we checking `WT_CURSTD_BOUND_LOWER_INCLUSIVE` here?

File: src/cursor/cur_std.c [hunk 5]
```diff
@@ -1175,13 +1176,83 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
+                if (exact < 0)
+                    WT_ERR_MSG(session, EINVAL,
+                      "The upper bounds is not lexicographically greater than the lower bound");
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1196]: Here I would just say "The upper bound overlaps with the lower bound"

File: src/cursor/cur_std.c [hunk 6]
```diff
@@ -1175,13 +1176,83 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+            }
+            F_SET(cursor, WT_CURSTD_BOUND_UPPER | WT_CURSTD_BOUND_UPPER_INCLUSIVE);
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1198]: Seems like we're not handling inclusive yet?

File: src/cursor/cur_std.c (hunk 1 above)
Comment (luke-pearson) [reply src/cursor/cur_std.c]: I don't think we want to call WT_CLEAR here. the key passed in by the user is being copied by us, we should clear the cursor key but there should a better mechanism to it.

Comment (luke-pearson) [reply src/cursor/cur_std.c]: You know what I don't think we need to clear at all, as @ddanderson said.

File: src/cursor/cur_std.c [hunk 7]
```diff
@@ -1175,13 +1179,88 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_UPPER)) {
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->upper_bound, &exact));
+                if (exact > 0)
+                    WT_ERR_MSG(session, EINVAL, "The lower bound overlaps with the upper bound");
+            }
+            /* Copy the key over to the upper bound item and set upper bound and inclusive flags. */
+            F_SET(cursor, WT_CURSTD_BOUND_LOWER);
+            WT_ERR(__wt_config_gets(session, cfg, "inclusive", &cval));
+            if (cval.val)
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1225]: Can you check in GDB that by default inclusive is set to true and we pick it up as true here?

**Review by ddanderson — COMMENTED** [review]
Mostly good, but I spotted one problem.  Test!

File: test/suite/test_cursor_bound02.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        cursor.set_key(self.gen_key(99))
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda: cursor.bound("bound=lower"), '/Invalid argument/')
+
+        # Test bound API: Test resetting lower bound to 20, which would succeed setting upper
+        # bound to 30
+        cursor.set_key(self.gen_key(20))
+        self.assertEqual(cursor.bound("bound=lower"), 0)
+        cursor.set_key(self.gen_key(30))
+        self.assertEqual(cursor.bound("bound=upper"), 0)
+
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply test/suite/test_cursor_bound02.py:96]: Sorry If I misinterpreted, but currently we don't have any implementations done that restricts the keys to be returned from any of the cursor operations. @ddanderson Was this the type of tests that you were looking for?

File: src/cursor/cur_std.c [hunk 9]
```diff
@@ -1175,13 +1179,88 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
+                if (exact < 0)
+                    WT_ERR_MSG(session, EINVAL, "The upper bound overlaps with the lower bound");
+            }
+            /* Copy the key over to the upper bound item and set upper bound and inclusive flags. */
+            F_SET(cursor, WT_CURSTD_BOUND_UPPER);
+            WT_ERR(__wt_config_gets(session, cfg, "inclusive", &cval));
+            if (cval.val)
+                F_SET(cursor, WT_CURSTD_BOUND_LOWER_INCLUSIVE);
```
[the remark this answers is not in the document]
Comment (jiechenbo) [reply src/cursor/cur_std.c:1203]: Currently there I can't think of a way to test this, since cursor operations aren't implemented yet, do you have an idea @ddanderson ?

File: src/cursor/cur_std.c [hunk 10]
```diff
@@ -1175,13 +1179,78 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+             * bound.
+             */
+            WT_ERR(__wt_cursor_get_raw_key(cursor, &key));
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_UPPER)) {
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->upper_bound, &exact));
+                if (exact > 0)
+                    WT_ERR_MSG(session, EINVAL, "The lower bound overlaps with the upper bound");
+            }
+            /* Copy the key over to the upper bound item and set upper bound and inclusive flags. */
```
Comment (clarissecheah) [inline src/cursor/cur_std.c:1217]: Did this mean to say "copy the key over to the lower bound item..." instead?

**Review by clarissecheah — CHANGES_REQUESTED** [review]

File: test/suite/test_cursor_bound02.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        self.assertEqual(cursor.bound("bound=lower"), 0)
+        cursor.set_key(self.gen_key(99))
+        self.assertEqual(cursor.bound("bound=upper"), 0)
+
+        # Test bound API: No key set.
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda: cursor.bound("bound=lower"), '/Invalid argument/')
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda: cursor.bound("bound=upper"), '/Invalid argument/')
+        
+        # Test bound API: Test that reset also clears both of the bounds. 
+        cursor.reset()
```
Comment (clarissecheah) [inline test/suite/test_cursor_bound02.py:125]: What happens when you set a new upper bound thats < the previous upper bound, but you don't do a reset. Is that expected to be valid behaviour? If so, does this actually test that reset() works as you'd be expected to be able to do this anyway.

File: src/btree/bt_cursor.c [hunk 12]
```diff
@@ -514,6 +514,16 @@ __wt_btcur_reset(WT_CURSOR_BTREE *cbt)
 
     WT_STAT_CONN_DATA_INCR(session, cursor_reset);
 
+    /* Clear bounds if they are set. */
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:517]: This seems like the right location to me but I've created a ticket to follow up later in the project, there are a few other potential functions for it so I want to make sure its in the right place. Theres no additional work right now here.

File: src/cursor/cur_std.c [hunk 13]
```diff
@@ -1175,13 +1179,78 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
         if (cval.len == 0)
             WT_ERR_MSG(session, EINVAL, "setting bounds must require the bound configuration set");
+
+        /* The cursor must have a key set to place the lower or upper bound. */
+        WT_ERR(__cursor_checkkey(cursor));
+
+        if (WT_STRING_MATCH("upper", cval.str, cval.len)) {
+            /*
+             * If the lower bounds are set, make sure that the upper bound is greater than the
+             lower
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1189]: Comment formatting is wrong here.

File: src/cursor/cur_std.c (hunk 7 above)
Comment (jiechenbo) [reply src/cursor/cur_std.c:1220]: Yep! Double checked that cval.val returns back 1 here
```
1219                WT_ERR(__wt_config_gets(session, cfg, "inclusive", &cval));
(gdb) 
1220                if (cval.val)
(gdb) p cval.val
$1 = 1
```

File: src/cursor/cur_std.c [hunk 14]
```diff
@@ -1175,13 +1179,78 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+             * bound.
+             */
+            WT_ERR(__wt_cursor_get_raw_key(cursor, &key));
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_LOWER)) {
+                WT_ERR(__wt_compare(
+                  session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
+                if (exact < 0)
+                    WT_ERR_MSG(session, EINVAL, "The upper bound overlaps with the lower bound");
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1197]: I would make sure the error message is the same for both cases that way its easier to handle at the MongoDB layer. "The provided cursor bounds are overlapping."

File: test/suite/test_cursor_bound02.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        # Test bound API: No key set.
+        cursor.reset()
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda: cursor.bound("bound=lower"), '/Invalid argument/')
+        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda: cursor.bound("bound=upper"), '/Invalid argument/')
+        
+        # Test bound API: Test that reset also clears both of the bounds. 
+        cursor.reset()
+        cursor.set_key(self.gen_key(80))
+        self.assertEqual(cursor.bound("bound=upper"), 0)
+        cursor.set_key(self.gen_key(30))
```
Comment (luke-pearson) [inline test/suite/test_cursor_bound02.py:129]: Can you add some additional cases:

1. A clear followed by a reset
2. Moving bounds, so setting a bound and then changing it.
3. Testing that the key persists, so do a set_key("blah) cursor->bound, followed by a set_value cursor->insert.

File: test/suite/test_cursor_bound02.py (hunk 11 above)
Comment (jiechenbo) [reply test/suite/test_cursor_bound02.py:126]: >What happens when you set a new upper bound thats < the previous upper bound, but you don't do a reset. Is that expected to be valid behaviour? 
Yep that should be total valid behaviour
>If so, does this actually test that reset() works as you'd be expected to be able to do this anyway.
In this case, we are checking if the lower bounds were cleared. Previously before the clearing the lower bounds were 90, thus placing upper bounds at 80 should return EINVAL, if the lower bound wasn't cleared.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: test/suite/test_cursor_bound02.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply test/suite/test_cursor_bound02.py]: Ah, sorry - I see you're doing it in stages.

File: src/cursor/cur_std.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/cursor/cur_std.c]: No, I don't have a better plan.  You can only test what you have implemented.

File: src/cursor/cur_std.c [hunk 16]
```diff
@@ -1175,13 +1179,76 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
...
+         * bounds, both the upper and lower bound will be cleared.
+         */
+        WT_ERR(__wt_config_gets(session, cfg, "bound", &cval));
+        if (cval.len == 0) {
+            F_CLR(cursor, WT_CURSTD_BOUNDS_ALL);
+            __wt_buf_free(session, &cursor->upper_bound);
+            WT_CLEAR(cursor->upper_bound);
+            __wt_buf_free(session, &cursor->lower_bound);
+            WT_CLEAR(cursor->lower_bound);
+        } else if (WT_STRING_MATCH("upper", cval.str, cval.len)) {
```
Comment (ddanderson) [inline src/cursor/cur_std.c:1242]: There's a way to remove repeated code here.  Remove the first clause (`if (cval.len == 0)`) entirely.  Then:
```
  if (cval.len == 0 || WT_STRING_MATCH("upper", cval.str, cval.len)) {
     ...
  }
  if (cval.len == 0 || WT_STRING_MATCH("lower", cval.str, cval.len)) {
     ...
  }
```
(No else clause).  Feel free to ignore if you don't like that approach.

**Review by ddanderson — APPROVED** [review]
LGTM - left a minor refactor suggestion, it's optional.

File: test/suite/test_cursor_bound02.py (hunk 11 above)
Comment (clarissecheah) [reply test/suite/test_cursor_bound02.py:258]: I misread this, this makes sense to me.

**Review by clarissecheah — APPROVED** [review]
LGTM!

**Comment by jiechenbo** [comment]
Fixed a problem in cppsuite where the std::string variable would go out of scope, and thus accessing weird memory.

File: src/cursor/cur_std.c [hunk 17]
```diff
@@ -1194,12 +1202,17 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
                   session, CUR2BT(cursor)->collator, &key, &cursor->lower_bound, &exact));
                 if (exact < 0)
                     WT_ERR_MSG(session, EINVAL, "The provided cursor bounds are overlapping");
+                /* If lower bound and upper bound is equal, both inclusive flags must be set. */
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1205]: Reword: "If the lower bound and upper bound are equal, both inclusive flags must be specified. ".

File: src/cursor/cur_std.c [hunk 18]
```diff
@@ -1212,12 +1225,17 @@ __wt_cursor_bound(WT_CURSOR *cursor, const char *config)
                   session, CUR2BT(cursor)->collator, &key, &cursor->upper_bound, &exact));
                 if (exact > 0)
                     WT_ERR_MSG(session, EINVAL, "The provided cursor bounds are overlapping");
+                /* If lower bound and upper bound is equal, both inclusive flags must be set. */
```
Comment (luke-pearson) [inline src/cursor/cur_std.c:1228]: Same here.

**Review by clarissecheah — APPROVED** [review]
